mesh/mesh-io-tcpserver.c

```diff
@@ -0,0 +1,637 @@
+#include "mesh/silvair-io.h"
+
+
+struct mesh_io_private {
+	struct l_io		*server_io;
+	struct l_queue		*client_io;
+
+	struct l_timeout	*tx_timeout;
+	struct l_queue		*rx_regs;
+	struct l_queue		*tx_pkts;
+
+	/* Simple filtering on AD type only */
+	uint8_t 		filters[3];
+	struct tx_pkt 		*tx;
+};
+
+struct pvt_rx_reg {
+	uint8_t filter_id;
+	mesh_io_recv_func_t cb;
+	void *user_data;
+};
+
+struct process_data {
+	struct mesh_io_private		*pvt;
+	const uint8_t			*data;
+	uint8_t				len;
+	struct mesh_io_recv_info	info;
+};
+
+struct tx_pkt {
+	uint32_t	instant;
+	uint8_t		len;
+	uint8_t		data[30];
+};
+
+struct tx_pattern {
+	const uint8_t			*data;
+	uint8_t				len;
+};
+
+enum io_type {
+    IO_TYPE_SERVER,
+    IO_TYPE_CLIENT
+};
+
+static void send_timeout(struct l_timeout *timeout, void *user_data);
+static void keep_alive_error(struct l_timeout *timeout, void *user_data);
+
+
+static uint32_t get_instant(void)
+{
+	struct timeval tm;
+	uint32_t instant;
+
+	gettimeofday(&tm, NULL);
+	instant = tm.tv_sec * 1000;
+	instant += tm.tv_usec / 1000;
+
+	return instant;
+}
+
+static void process_rx_callbacks(void *v_rx, void *v_reg)
+{
+	struct pvt_rx_reg *rx_reg = v_rx;
+	struct process_data *rx = v_reg;
+	uint8_t ad_type;
+
+	ad_type = rx->pvt->filters[rx_reg->filter_id - 1];
+
+	if (rx->data[0] == ad_type && rx_reg->cb)
+		rx_reg->cb(rx_reg->user_data, &rx->info, rx->data, rx->len);
+}
+
+static void process_rx(struct silvair_io *silvair_io, int8_t rssi,
+			const uint8_t *data, uint8_t len, void *user_data)
+{
+	struct mesh_io *mesh_io = user_data;
+
+	struct process_data rx = {
+		.pvt = mesh_io->pvt,
+		.data = data,
+		.len = len,
+		.info.instant = get_instant(),
+		.info.chan = 7,
+		.info.rssi = rssi,
+	};
+
+	if (!mesh_io) {
+		l_error("mesh_io does not exist");
+		return;
+	}
+
+	silvair_io_keep_alive_wdt_refresh(silvair_io);
+	l_queue_foreach(mesh_io->pvt->rx_regs, process_rx_callbacks, &rx);
+}
+
+static bool get_fd_info(int fd, char *log, enum io_type type)
+{
+	/* Address */
+	struct sockaddr_in addr;
+	unsigned int addrlen;
+
+	addrlen = sizeof(addr);
+
+	switch(type) {
+		case IO_TYPE_CLIENT:
+			if (getpeername(fd,(struct sockaddr *)&addr, &addrlen) < 0) {
+				l_error("getpeername() error");
+				return false;
+			}
+			break;
+
+		case IO_TYPE_SERVER:
+			if (getsockname(fd,(struct sockaddr *)&addr, &addrlen) < 0) {
+				l_error("getsockname() error");
+				return false;
+			}
+			break;
+
+		default:
+			l_error("get_df_info() Invalid type");
+			return false;
+	}
+
+	l_info("%s -> addr:%s port:%d fd:%d", log,
+			inet_ntoa(addr.sin_addr), ntohs(addr.sin_port), fd);
+	return true;
+}
+
+static void io_read_callback_destroy(void *user_data)
+{
+	struct silvair_io *silvair_io = user_data;
+	int fd = l_io_get_fd(silvair_io->l_io);
+
+	get_fd_info(fd, "Disconnecting the TCP client", IO_TYPE_CLIENT);
+
+	/* close will trigger the io_disconnect_callback() */
+	close(fd);
+}
+
+static void io_disconnect_callback(void *user_data)
```

Comment:
I would move this function inside `silvair-io`, extract `->context` there and then call callback provided by `mesh_io` layer in `silvair_io` constructor.

The point is to make `silvaie_io` structure opaque from `mesh_io` point of view.